keyberon/src/layout.rs

```diff
@@ -3584,4 +3668,339 @@ mod test {
+
+        layout.event(Press(0, 0));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        layout.event(Press(0, 1));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        layout.event(Press(0, 2));
+        for _ in 0..10 {
+            assert_eq!(CustomEvent::NoEvent, layout.tick());
+            assert_keys(&[], layout.keycodes());
+        }
+        layout.event(Release(0, 2));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[C], layout.keycodes());
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+    }
+
+    #[test]
+    fn test_trans_in_fork() {
+        static LAYERS: Layers<3, 1, 3> = [
+            [[Layer(1), NoOp, k(A)]],
+            [[NoOp, Layer(2), k(B)]],
+            [[
+                NoOp,
+                NoOp,
+                Fork(&ForkConfig {
+                    left: Trans,
+                    right: Trans,
+                    right_triggers: &[Space],
+                }),
+            ]],
+        ];
+        let mut layout = Layout::new(&LAYERS);
+
+        layout.event(Press(0, 0));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        layout.event(Press(0, 1));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        layout.event(Press(0, 2));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[B], layout.keycodes());
+        layout.event(Release(0, 2));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+    }
+
+    #[test]
+    fn test_trans_in_switch() {
+        static LAYERS: Layers<3, 1, 3> = [
+            [[Layer(1), NoOp, k(A)]],
+            [[NoOp, Layer(2), k(B)]],
+            [[
+                NoOp,
+                NoOp,
+                Switch(&switch::Switch {
+                    cases: &[(&[], &Trans, BreakOrFallthrough::Break)],
+                }),
+            ]],
+        ];
+        let mut layout = Layout::new(&LAYERS);
+
+        layout.event(Press(0, 0));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        layout.event(Press(0, 1));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+
+        layout.event(Press(0, 2));
+        // No idea why we have to wait 2 ticks here. Is this a bug in switch?
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[B], layout.keycodes());
+
+        layout.event(Release(0, 2));
+        assert_eq!(CustomEvent::NoEvent, layout.tick());
+        assert_keys(&[], layout.keycodes());
+    }
```

Comment:
Here's a failing test:

<details>
<summary>test_multi_taphold_trans</summary>

```rust
    #[test]
    fn test_multi_taphold_trans() {
        static LAYERS: Layers<4, 1, 4> = [
            [[Layer(1), NoOp, NoOp, k(A)]],
            [[NoOp, Layer(2), NoOp,
                HoldTap(&HoldTapAction {
                    timeout: 50,
                    hold: k(B),
                    timeout_action: k(B),
                    tap: Trans,
                    config: HoldTapConfig::Default,
                    tap_hold_interval: 200,
                }),
            ]],
            [[
                NoOp,
                NoOp,
                Layer(3),
                HoldTap(&HoldTapAction {
                    timeout: 50,
                    hold: k(C),
                    timeout_action: k(C),
                    tap: Trans,
                    config: HoldTapConfig::Default,
                    tap_hold_interval: 200,
                }),
            ]],
            [[
                NoOp,
                NoOp,
                Layer(3),
                HoldTap(&HoldTapAction {
                    timeout: 50,
                    hold: k(D),
                    timeout_action: k(D),
                    tap: Trans,
                    config: HoldTapConfig::Default,
                    tap_hold_interval: 200,
                }),
            ]],
        ];
        let mut layout = Layout::new(&LAYERS);

        layout.event(Press(0, 0));
        assert_eq!(CustomEvent::NoEvent, layout.tick());
        assert_keys(&[], layout.keycodes());
        layout.event(Press(0, 1));
        assert_eq!(CustomEvent::NoEvent, layout.tick());
        assert_keys(&[], layout.keycodes());

        layout.event(Press(0, 2));
        assert_eq!(CustomEvent::NoEvent, layout.tick());
        assert_keys(&[], layout.keycodes());
        layout.event(Press(0, 3));
        assert_eq!(CustomEvent::NoEvent, layout.tick());
        assert_keys(&[], layout.keycodes());
        layout.event(Release(0, 3));
        assert_eq!(CustomEvent::NoEvent, layout.tick());
        assert_keys(&[A], layout.keycodes());
    }
```

</details>

I don't know that anyone would _intentionally_ do this, but since this feature enables adding multiple layers on top of each other, someone may unintentionally end up doing something like this and then it doesn't work as expected (and in fact seems like it ends up in an endless loop?).

The cause of this is that the `skip(1)` in `waiting_into_*` isn't quite correct.

Reply:
fixed in 7a2c0b95cd3cc2c7634739aa193c44a55f09839c